lib/OpenQA/Schema/Result/ScheduledProducts.pm

```diff
@@ -753,4 +727,122 @@ sub _create_download_lists {
+        next
+          if ( $product->{distri} ne _distri_key($args)
+            || $product->{flavor} ne $args->{FLAVOR}
+            || $product->{version} ne $args->{VERSION}
+            || $product->{arch} ne $args->{ARCH});
+        my $settings = $job_template->{settings} // {};
+        $settings->{TEST} = $key;
+        my @worker_class;
+        push @worker_class, $settings->{WORKER_CLASS} if $settings->{WORKER_CLASS};
+
+        my $product_settings = $product->{settings} // {};
+        foreach (keys %$product) {
+            $settings->{uc $_} = $product->{$_} if $_ ne 'settings';
+        }
+        foreach my $s_key (keys %$product_settings) {
+            if ($s_key eq 'WORKER_CLASS') {
+                push @worker_class, $product_settings->{$s_key};
+                next;
+            }
+            $settings->{$s_key} = $product_settings->{$s_key};
+        }
+        if (my $machine = $job_template->{machine}) {
+            $settings->{MACHINE} = $machine;
+            if (my $mach = $machines->{$machine}) {
+                my $machine_settings = $mach->{settings} // {};
+                foreach (keys %$machine_settings) {
+                    if ($_ eq 'WORKER_CLASS') {
+                        push @worker_class, $machine_settings->{WORKER_CLASS};
+                        next;
+                    }
+                    $settings->{$_} = $machine_settings->{$_};
+                }
+                $settings->{BACKEND} = $mach->{backend} if $mach->{backend};
+                $settings->{PRIO} = $mach->{priority} // 50;
+            }
+        }
+        $settings->{WORKER_CLASS} = join ',', sort @worker_class if @worker_class > 0;
+        foreach (keys %$args) {
+            $settings->{uc $_} = $args->{$_} if $_ ne 'TEST';
+        }
+        $settings->{DISTRI} = lc($settings->{DISTRI}) if $settings->{DISTRI};
+
+        OpenQA::JobSettings::parse_url_settings($settings);
+        OpenQA::JobSettings::handle_plus_in_settings($settings);
+        my $error = OpenQA::JobSettings::expand_placeholders($settings);
+        $error_msg .= $error if defined $error;
+
+        _count_wanted_jobs($args, $settings, \%wanted);
+        push @job_templates, $settings;
+    }
+
+    return {
+        settings_result => _handle_dependency(\@job_templates, \%wanted, $skip_chained_deps),
+        error_message => $error_msg,
+    };
+}
+
+sub _count_wanted_jobs ($args, $settings, $wanted) {
+    my @tests = $args->{TEST} ? split(/\s*,\s*/, $args->{TEST}) : ();
+    if (!$args->{MACHINE} || $args->{MACHINE} eq $settings->{MACHINE}) {
+        if (!@tests) {
+            $wanted->{_settings_key($settings)} = 1;
+        }
+        else {
+            foreach my $test (@tests) {
+                if ($test eq $settings->{TEST}) {
+                    $wanted->{_settings_key($settings)} = 1;
+                    last;
+                }
+            }
+        }
+    }
+}
+
+sub _handle_dependency ($jobs, $wanted, $skip_chained_deps) {
```

Comment:
This function could also have a better name and maybe also deserves better comments (to better convey its function in the bigger picture).  (The function is self is of course nothing new. The code has simply been extracted to avoid duplication.)